pkg/apis/networking/types.go

```diff
@@ -544,3 +544,68 @@ type ServiceBackendPort struct {
 	// +optional
 	Number int32
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterCIDRConfig is the Schema for the clustercidrconfigs API.
+type ClusterCIDRConfig struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   ClusterCIDRConfigSpec
+	Status ClusterCIDRConfigStatus
+}
+
+// ClusterCIDRConfigSpec defines the desired state of ClusterCIDRConfig.
+type ClusterCIDRConfigSpec struct {
+	// NodeSelector defines which nodes the config is applicable to.
+	// An empty or nil NodeSelector functions as a default that applies to all nodes.
+	// This field is immutable.
+	// +optional
+	NodeSelector *api.NodeSelector
+
+	// IPv4 defines the IPv4 CIDR and the PerNodeMaskSize.
+	// At least one of IPv4 or IPv6 must be provided. If both are
+	// provided, the number of IPs allocated to each must be the same
+	// (32 - ipv4.perNodeMaskSize).
+	// This field is immutable.
+	// +optional
+	IPv4 *CIDRConfig
+
+	// IPv6 defines the IPv4 CIDR and the PerNodeMaskSize.
+	// At least one of IPv4 or IPv6 must be provided. If both are
+	// provided, the number of IPs allocated to each must be the same
+	// (128 - ipv6.perNodeMaskSize).
+	// This field is immutable.
+	// +optional
+	IPv6 *CIDRConfig
+}
+
+// CIDRConfig defines the CIDR and Mask size per IP Family(IPv4/IPv6).
+type CIDRConfig struct {
+	// An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64").
+	CIDR string
+
+	// PerNodeMaskSize is the mask size for node cidr.
+	// IPv4/IPv6 Netmask size (e.g. 25 -> "/25" or 112 -> "/112") to allocate to a node.
+	PerNodeMaskSize int32
+}
+
+// ClusterCIDRConfigStatus defines the observed state of ClusterCIDRConfig.
+type ClusterCIDRConfigStatus struct {
+}
```

Comment:
I understand that this is not part of the original kep. But may worth considering updating usage statistic e.g, % used with a timestamp. I don't want us to end up with every alloc updates the usage statistics but we can do that on 1m/5m interval.